src/node/app/proxy.ts

```diff
@@ -0,0 +1,190 @@
+import { logger } from "@coder/logger"
+import * as http from "http"
+import proxy from "http-proxy"
+import * as net from "net"
+import * as querystring from "querystring"
+import { HttpCode, HttpError } from "../../common/http"
+import { HttpProvider, HttpProviderOptions, HttpProxyProvider, HttpResponse, Route } from "../http"
+
+interface Request extends http.IncomingMessage {
+  base?: string
+}
+
+/**
+ * Proxy HTTP provider.
+ */
+export class ProxyHttpProvider extends HttpProvider implements HttpProxyProvider {
+  /**
+   * Proxy domains are stored here without the leading `*.`
+   */
+  public readonly proxyDomains: string[]
+  private readonly proxy = proxy.createProxyServer({})
+
+  /**
+   * Domains can be provided in the form `coder.com` or `*.coder.com`. Either
+   * way, `<number>.coder.com` will be proxied to `number`.
+   */
+  public constructor(options: HttpProviderOptions, proxyDomains: string[] = []) {
+    super(options)
+    this.proxyDomains = proxyDomains.map((d) => d.replace(/^\*\./, "")).filter((d, i, arr) => arr.indexOf(d) === i)
+    this.proxy.on("error", (error) => logger.warn(error.message))
+    // Intercept the response to rewrite absolute redirects against the base path.
+    this.proxy.on("proxyRes", (response, request: Request) => {
+      if (response.headers.location && response.headers.location.startsWith("/") && request.base) {
+        response.headers.location = request.base + response.headers.location
+      }
+    })
+  }
+
+  public async handleRequest(
+    route: Route,
+    request: http.IncomingMessage,
+    response: http.ServerResponse,
+  ): Promise<HttpResponse> {
+    const isRoot = !route.requestPath || route.requestPath === "/index.html"
+    if (!this.authenticated(request)) {
+      // Only redirect from the root. Other requests get an unauthorized error.
+      if (isRoot) {
+        return { redirect: "/login", query: { to: route.fullPath } }
+      }
+      throw new HttpError("Unauthorized", HttpCode.Unauthorized)
+    }
+
+    // Ensure there is a trailing slash so relative paths work correctly.
+    const port = route.base.replace(/^\//, "")
+    const base = `${this.options.base}/${port}`
+    if (isRoot && !route.fullPath.endsWith("/")) {
+      return {
+        redirect: `${base}/`,
+      }
+    }
+
+    const payload = this.doProxy(route, request, response, port, base)
```

Comment:
Why does this take a response but also return a payload?

Reply:
Yeah I don't really like this either. `http-proxy` takes the response and deals with it by itself which is what the response is needed for. But the proxy might want to do a redirection or something in which case it needs to return a payload.

Oh maybe we can just return proxy information as part of the payload instead. Essentially moving `doProxy` out of `proxy.ts` and into `http.ts`.

Reply:
I like that.